### PR TITLE
Templates: swap templates to screenshots

### LINKED ISSUE #6678: Page Templates: Clicking on a template after applying a different one causes the thumbnails to jump
## Bug Description

If I scroll through the Page Layouts panel and apply one page, confirm and then scroll through it again and click on a different page layout, the panel automatically jumps to a different position, without applying the layout I clicked on.

## Expected Behaviour

Panel should not automatically jump, and user should be able to click any page layout visible on the panel at any moment to apply it.

## Steps to Reproduce

Apply a page layout, confirm, scroll through panel and click on a different page layout

## Screenshots

https://images.zenhubusercontent.com/235435637/72a65a24-703d-4e77-9ae5-b37c1b261477/screen_recording_2021_03_09_at_11_06_41_am.mp4

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System: MacOS
- Browser: Chrome Version 88.0.4324.192 (Official Build) (x86_64)

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### LINKED ISSUE #9093: Default Page Templates: Swap for screenshots
## Description

This will include: removing the virtualized list, making sure page templates are memoized so we don't load excessively, and making sure that filtering and toggling images are still working - essentially, ensuring no regression. Right now the custom and default page templates share some components, it would probably make sense at this point to make a new page templates list component for static page templates, so that this work doesn't impact custom page templates at all but could be used for those as well once looking into that. 

## To Do 
- Do we want to keep the placeholder grid around? (@choumx following up on this) 
- Do we want to swap the toggle so that images are there by default? (@choumx following up on this)

Reference: #8826

## Acceptance Criteria

- Using screenshots for default page templates instead of previewing stories for each page. 
- Filtering by page template type still works. 
- Toggling between templates with images and with the placeholder has been removed
- Performance metrics to see improvements. 
- Passing tests
- Added aXe integration test for accessibility -- will be done in #8736

### PR DESCRIPTION
**Pull request by BrookeGraham**
## Summary

Using screenshots for default page templates instead of previewing stories for each page.
Filtering by page template type still works.
Toggling between templates with images and with the placeholder grid removed

Old metrics for initial load: 

staging load 1
staging load 2
staging load 3

New metrics for initial load:

new load 1
new load 2
new load 3

---
Old metrics for scrolling to the end of templates list:
staging scrolling 1
staging scrolling 2
staging scrolling 3

New metrics for scrolling to the end of templates list:
new scrolling 1
new scrolling 2
new scrolling 3

## To-do

- [x] Use arrow for keyboard nav. 
- [x] Clicking to add page bug, right now you have to click twice on an image before `handlePageClick` fires 🤷🏻‍♀️
- [x] Tests - aXe integration test for accessibility. - using useGridViewKeys, which have tests for keyboard accessibility.
- [x] Performance metrics to see improvements. 

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Make sure default and saved templates work same as before. 
2. Should be keyboard navigable
3. Default templates should now appear as images

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?
no

### Does this PR have a legal-related impact?
no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #9093 
Fixes #6678

### CODE REVIEW

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrookeGraham) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: This is a little hacky, totally open to thoughts on how to allow users to exit the focus trap.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Now that these are getting the default DS button styles focus styles are taken care of already so `focusStyle` is not necessary. 
```suggestion
```

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Not your fault but we should look at fixing[ this contrast ratio](https://contrast-ratio.com/#%23E4E5E6-on-rgba%280%2C0%2C0%2C0.47%29), it wasn't noticeable before to me but now it is:

Screen Shot 2021-10-07 at 4 35 41 PM

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: 🤔 Perhaps this would be better splitting the container and the text now that the DOM is so much cleaner? 

That would make it really straight forward to use `Text` from the design system, like so: 
`<Text as="span" size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.SMALL} isBold />`

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Now that nothing is animating, can we rely on browser level hover focus?

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: nit: weird line break? 

```suggestion
                /* translators: 1: template name. 2: page template name. */
                _x('%1$s %2$s', 'page template title', 'web-stories'),
```

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Let's use css pseudo classes here instead of `isActive`

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: You're passing rest in here as well as in the wrapper, is it needed in both (or either)?

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: Took this for a little spin and think we can simplify it a bit :) 

`useGridViewKeys` should be taking care of all of this now that it's just a group of buttons that we want to traverse with arrow keys! I'm leaving comments through here to show what I'm thinking.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: This can be deleted, focus should be handled by the `useGridViewKeys` util.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: This can also be deleted, as focus should be handled by that `useGridViewKeys` util as well.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: We don't need any of this either now that we're just dealing with buttons! A lot of the code that was in here was the way it was because of how gross the HTML was for rendering actual story previews.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: These can be removed. Let's use the `currentPageId` in tandem with the `page.id` when mapping through the list to find out which button should have tabIndex 0.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: Let's remove this onFocus and pass in `onFocus` to the `DefaultPageTemplate` since buttons are organically focusable.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: in additoin to setting the currentPageId here, let's add a `useEffect` above doing something like this: 
```
if (pages.length > 0 && !currentPageId) {
      setCurrentPageId(pages[0].id);
    }
```
so that if (first time) the user is tabbing to this grid, the first button is set to have focus available since `currentPageId` is getting used to decide `tabIndex`

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: remove :)

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: These can hopefully be removed

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: So part of the issue while trying to use the `useGridViewKeys` util that you were running up against was the button being the "wrapper" here, take a look at other implementations of this hook and see that there's a div around the focusable item. If you wrap `PageTemplateWrapper` (maybe renaming it `PageTemplateButton`?) in a `<div>` and put the `ref` on the `div` and everything else on the button (ask it is) and also add in: `tabIndex={isActive ? 0 : -1}` focusing should work.

**Review by BrittanyIRL — CHANGES_REQUESTED** [review]
😅  this was a _lot_ to untangle and refactor, great work. 

Left some comments about further clean up and simplification. Happy to jump on a call next week about anything!!

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: I left this as a div since we have so much custom styling that comes with it. ie. focus and hover state, absolute positioning, etc.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Should I check with Amy or Andrew on this?

Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Asked in design channel. Will update once I receive confirmation.

Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Any confirmed that she's okay with using the darker theme color.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (maxyinger) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: I would just break up the styling per element a little bit to use some of our text styling. just make this `PageTemplateTitleContainer` and do something like:
```
            <PageTemplateTitleContainer>
              <Text as="span" size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.SMALL} isBold >
                {sprintf(
                  /* translators: 1: template name. 2: page template name. */
                  _x('%1$s %2$s', 'page template title', 'web-stories'),
                  page.title,
                  PAGE_TEMPLATE_TYPES[page.type].name
                )}
              <Text />
            </PageTemplateTitleContainer>
```

either that or use one of the `themeHelpers` like `themeHelpers.expandPresetStyles({ preset, theme })` to get the majority of the font styles in here, then tweak as needed with custom styles

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (maxyinger) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Adding draggable false here should prevent the user from being able to drag the image around and possibly making them thing that this component is draggable
```jsx
              <PosterImg
                src={page.png}
                alt={page.title}
                crossOrigin="anonymous"
                draggable={false}
              />
```

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (maxyinger) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Okay, I think we can clean these styles up a bit. 

I think the mental model we should go by here is that the parent container sets the grid column width, so we should just respect that width all the way down to the image and make the image take up that full width. The image should then layout according to its own aspect ratio which should set the height for everything. I'll walk you through the css with some comments sprinkled in this file.

Comment (maxyinger) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Here are the properties we'll remove
```css
  /* the contents of this element are 1 element taking up the full size and another element 
      laid over top of it, so we don't need these properties*/
  display: flex;
  flex-direction: column;
  
  /* we're gonna let the image dictate it's own height later, so we don't need to worry about it here */
  height: auto;
  
  /* we're just taking up the full width of the parent component, so we don't need an explicit width here */
  width: ${({ columnWidth }) => columnWidth}px;
  
  /* this should already be inherited by the base button class, so we don't need to redeclare it here */
  cursor: pointer;
```

```suggestion
const PageTemplateButton = styled(Button).attrs({ type: BUTTON_TYPES.PLAIN })`
 /* will take up full width of parent */
  display: block;
  
  /* this will make all relative & absolute positioned elements relative to this */
  position: relative;
  
  padding: 0;
  border-radius: ${({ theme }) => theme.borders.radius.small};
  
   /* since the overflow is hidden, we won't need to redeclare border-radiuses on children cus those bits will just get cut off with this */
  overflow: hidden;
`;
```

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (maxyinger) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]:
```suggestion
  /* since the button above is relative, this will lay this content over all the other elements in here and position it at the bottom. no need to touch anything regarding flex here*/
  position: absolute;
  bottom: 0;
  
  width: 100%;
  padding: 8px;
  background-color: ${({ theme }) => theme.colors.opacity.black64};
  opacity: 0;

  ${PageTemplateButton}:hover &,
  ${PageTemplateButton}:focus & {
    opacity: 1;
  }
```

Comment (maxyinger) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: since the parent has `overflow: hidden` we don't have to worry about rounding any edges here because all the edges are already visually cut off by the parent

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (maxyinger) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: We can just remove this container entirely and position the image directly inside the button

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (maxyinger) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]:
```suggestion
const PosterImg = styled.img`
  display: block;
   /* by default the image lays out to it's true width, we want to manually set it to take up the full width of it's parent */
   width: 100%; 
`;
```

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js
Comment (maxyinger) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js:65]: ahh so if we do want the template images can just pass a simple showImages here:
```
      setPageTemplates(
        await getPageTemplates({ showImages: true })
      );
```

File: packages/story-editor/src/components/library/panes/pageTemplates/templateList.js
Comment (maxyinger) [inline packages/story-editor/src/components/library/panes/pageTemplates/templateList.js]: should we still be cautionary here?
```
 const pageIds = useMemo(() => pages?.map((page) => page.id) || [], [pages]);
```

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js
Comment (samwhale) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js:185]: Just to double check: we're removing these because we're now rendering these as static images. So we can't get screenshots of what they look like without the images?

**Review by maxyinger — APPROVED** [review]
Looks great! Just left some minor implementation details that don't really change anything here. Awesome work!! so much cleaner and works so much better 👏

**Review by samwhale — APPROVED** [review]
Nice, max's review covers css stuff in depth. I just have one question about a toggle that was removed, but other than that it works great

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js:185]: Correct.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js:65]: For now, we want to add templates without images.  [See thread](https://formidablelabs.slack.com/archives/C01ESBB8ZN1/p1634236997169600?thread_ts=1634234530.168300&cid=C01ESBB8ZN1)

**Comment by BrookeGraham** [comment]
> Looks great! Just left some minor implementation details that don't really change anything here. Awesome work!! so much cleaner and works so much better 👏

Thanks, so much better now!

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js
Comment (BrittanyIRL) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js:65]: this still feels weird. i get that this came from design but it feels odd to me as a user to select a template and then have it show up and look different from what i expected.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: nit: you're looking for `page.webp` here and then using `page.png` as the img src within, should probably check for `.png` if using that or swap to `webp`

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: `PageTemplateTitleContainer` is getting `isActive` but that prop's not getting passed through at all. If it's still needed it should be a transient prop and passed through like `<PageTemplateTitleContainer $isActive={isActive}` since it's only getting passed for styles and serves no other purpose at that level.  

That said, pretty sure with the simplifications you were able to make this is obsolete! 😅

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: 🤔  Callback?
`onFocus={() => handlePageFocus(page.id)}`

`const handlePageFocus = ({id}) => setCurrentPageId(id), []);`

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js]: Was there a specific purpose in setting this here separately from the spread `poster` below it?

**Review by BrittanyIRL — CHANGES_REQUESTED** [review]
Thank you for the updates! 

Few smaller things I noticed on re-review and then these two things that are more generic: 
Screen Shot 2021-10-18 at 11 55 05 AM

Looks like on scroll the templates are overlapping with this button. Ideally, there's no overlap.
I know this isn't necessarily new, but we should leave it better off with this update since the styles are getting a facelift in this PR. 

The focusing is _much_ improved, i did notice one thing though that i'm curious if you can reproduce. If you use the keyboard and filter templates to anything besides all and then go into the grid and select one, then tab back to the filters and select a new filter that isn't all and tab back to the grid - the focus disappears.

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplates.js]: nope. Removed

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultTemplateList.js]: ☑️

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Nice catch!

File: packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/library/panes/pageTemplates/defaultPageTemplate.js]: Totes.

**Comment by BrookeGraham** [comment]
> Looks like on scroll the templates are overlapping with this button. Ideally, there's no overlap.
I know this isn't necessarily new, but we should leave it better off with this update since the styles are getting a facelift in this PR.

Fixed spacing

> The focusing is much improved, i did notice one thing though that i'm curious if you can reproduce. If you use the keyboard and filter templates to anything besides all and then go into the grid and select one, then tab back to the filters and select a new filter that isn't all and tab back to the grid - the focus disappears.
 
Found it. The `currentPageId` was being set to the first item on initial load, which was a template type `cover` so when we filtered to a type that wasn't `cover`, the template was no longer `active`. Added a check in the useEffect to make sure the `currentPageId` is indeed in the pages array after filtering.

**Review by BrittanyIRL — APPROVED** [review]
This works great and is truly a massive improvement, thank you for all of your diligence here.

**Comment by BrittanyIRL** [comment]
(I think the page template karma tests are going to need some updates - i know there's a follow up for that but we can't merge this (after qa) with the failing tests)

**Comment by BrookeGraham** [comment]
> (I think the page template karma tests are going to need some updates - i know there's a follow up for that but we can't merge this (after qa) with the failing tests)

Yeah totally, they are failing right now. I will update the tests and link the ticket to this pr.

**Comment by swissspidy** [comment]
Is this e2e test really flakey? 🤨

It seems to run fine on main, but is broken in this PR, indicating a regression that should get fixed, not skipped.

**Pull request merged**
